pkg/controllermanager/controller/shoot/shoot_maintenance_control.go

```diff
@@ -378,6 +378,31 @@ func hasMaintainNowAnnotation(shoot *gardencorev1beta1.Shoot) bool {
 	return ok && operation == v1beta1constants.ShootOperationMaintain
 }
 
+func needsRetry(shoot *gardencorev1beta1.Shoot) bool {
+	needsRetryOperation := false
+
+	if val, ok := shoot.Annotations[v1beta1constants.FailedShootNeedsRetryOperation]; ok {
+		needsRetryOperation, _ = strconv.ParseBool(val)
+	}
+	delete(shoot.Annotations, v1beta1constants.FailedShootNeedsRetryOperation)
+
+	return needsRetryOperation
+}
+
+func getOperation(shoot *gardencorev1beta1.Shoot) string {
+	var (
+		operation            = v1beta1constants.GardenerOperationReconcile
+		maintenanceOperation = shoot.Annotations[v1beta1constants.GardenerMaintenanceOperation]
+	)
+
+	if isValid, _ := gardencorev1beta1helper.IsValidShootOperation(maintenanceOperation, shoot.Status.LastOperation); isValid {
+		operation = maintenanceOperation
+	}
+	delete(shoot.Annotations, v1beta1constants.GardenerMaintenanceOperation)
```

Comment:
It should be a side-effect as well. Probably we should rename the func or move the deletion outside.